--- a/modules/lsp-dsp-lib/include/private/dsp/arch/x86/avx2/pmath/log.h
+++ b/modules/lsp-dsp-lib/include/private/dsp/arch/x86/avx2/pmath/log.h
@@ -46,17 +46,17 @@
 
             static const float LOGB_C[] __lsp_aligned32 =
             {
-                LSP_DSP_VEC8(2.0f * M_LOG2E)
+                LSP_DSP_VEC8(2.0f * (float)M_LOG2E)
             };
 
             static const float LOGE_C[] __lsp_aligned32 =
             {
-                LSP_DSP_VEC8(1.0f / M_LOG2E)
+                LSP_DSP_VEC8(1.0f / (float)M_LOG2E)
             };
 
             static const float LOGD_C[] __lsp_aligned32 =
             {
-                LSP_DSP_VEC8(2.0f * M_LOG10E),
+                LSP_DSP_VEC8(2.0f * (float)M_LOG10E),
                 LSP_DSP_VEC8(0.301029995663981f) // 1/log2(10)
             };
         )
--- a/modules/lsp-dsp-lib/include/private/dsp/arch/x86/sse2/pmath/log.h
+++ b/modules/lsp-dsp-lib/include/private/dsp/arch/x86/sse2/pmath/log.h
@@ -46,17 +46,17 @@
 
             static const float LOGB_C[] __lsp_aligned16 =
             {
-                LSP_DSP_VEC4(2.0f * M_LOG2E)
+                LSP_DSP_VEC4(2.0f * (float)M_LOG2E)
             };
 
             static const float LOGE_C[] __lsp_aligned16 =
             {
-                LSP_DSP_VEC4(1.0f / M_LOG2E)
+                LSP_DSP_VEC4(1.0f / (float)M_LOG2E)
             };
 
             static const float LOGD_C[] __lsp_aligned16 =
             {
-                LSP_DSP_VEC4(2.0f * M_LOG10E),
+                LSP_DSP_VEC4(2.0f * (float)M_LOG10E),
                 LSP_DSP_VEC4(0.301029995663981f) // 1/log2(10)
             };
         )
--- a/modules/lsp-dsp-lib/include/private/dsp/arch/x86/avx512/pmath/log.h
+++ b/modules/lsp-dsp-lib/include/private/dsp/arch/x86/avx512/pmath/log.h
@@ -46,17 +46,17 @@
 
             static const float LOGB_C[] __lsp_aligned64 =
             {
-                LSP_DSP_VEC16(2.0f * M_LOG2E)
+                LSP_DSP_VEC16(2.0f * (float)M_LOG2E)
             };
 
             static const float LOGE_C[] __lsp_aligned64 =
             {
-                LSP_DSP_VEC16(1.0f / M_LOG2E)
+                LSP_DSP_VEC16(1.0f / (float)M_LOG2E)
             };
 
             static const float LOGD_C[] __lsp_aligned64 =
             {
-                LSP_DSP_VEC16(2.0f * M_LOG10E),
+                LSP_DSP_VEC16(2.0f * (float)M_LOG10E),
                 LSP_DSP_VEC16(0.301029995663981f) // 1/log2(10)
             };
         )
