--- a/src/modules/clock.cpp
+++ b/src/modules/clock.cpp
@@ -359,7 +359,7 @@
           }
 
           os << Glib::ustring::format((cldWPos_ != WS::LEFT || line == 0) ? std::left : std::right,
-                                      std::setfill(L' '),
+                                      std::setfill(' '),
                                       std::setw(cldMonColLen_ + ((line < 2) ? cldWnLen_ : 0)),
                                       getCalendarLine(today, ymTmp, line, firstdow, &m_locale_));
 
