--- a/grub-core/commands/date.c
+++ b/grub-core/commands/date.c
@@ -21,11 +21,18 @@
 #include <grub/err.h>
 #include <grub/misc.h>
 #include <grub/datetime.h>
-#include <grub/command.h>
+#include <grub/extcmd.h>
 #include <grub/i18n.h>
+#include <grub/term.h>
 
 GRUB_MOD_LICENSE ("GPLv3+");
 
+static const struct grub_arg_option options[] =
+  {
+    {0, 'n', 0, N_("Do not output the trailing newline."), 0, 0},
+    {0, 0, 0, 0, 0, 0}
+  };
+
 #define GRUB_DATETIME_SET_YEAR		1
 #define GRUB_DATETIME_SET_MONTH		2
 #define GRUB_DATETIME_SET_DAY		4
@@ -34,9 +41,10 @@ GRUB_MOD_LICENSE ("GPLv3+");
 #define GRUB_DATETIME_SET_SECOND	32
 
 static grub_err_t
-grub_cmd_date (grub_command_t cmd __attribute__ ((unused)),
+grub_cmd_date (grub_extcmd_context_t ctxt,
                int argc, char **args)
 {
+  struct grub_arg_list *state = ctxt->state;
   struct grub_datetime datetime;
   int limit[6][2] = {{1980, 2079}, {1, 12}, {1, 31}, {0, 23}, {0, 59}, {0, 59}};
   int value[6], mask;
@@ -46,10 +53,15 @@ grub_cmd_date (grub_command_t cmd __attribute__ ((unused)),
       if (grub_get_datetime (&datetime))
         return grub_errno;
 
-      grub_printf ("%d-%02d-%02d %02d:%02d:%02d %s\n",
+      grub_printf ("%d-%02d-%02d %02d:%02d:%02d %s %s",
                    datetime.year, datetime.month, datetime.day,
                    datetime.hour, datetime.minute, datetime.second,
-                   grub_get_weekday_name (&datetime));
+                   N_("UTC"), grub_get_weekday_name (&datetime));
+
+      if (!state[0].set)
+          grub_printf("\n");
+
+      grub_refresh ();
 
       return 0;
     }
@@ -130,17 +147,20 @@ fail:
   return grub_error (GRUB_ERR_BAD_ARGUMENT, "invalid datetime");
 }
 
-static grub_command_t cmd;
+static grub_extcmd_t cmd;
 
 GRUB_MOD_INIT(date)
 {
   cmd =
-    grub_register_command ("date", grub_cmd_date,
-			   N_("[[year-]month-day] [hour:minute[:second]]"),
-			   N_("Display/set current datetime."));
+    grub_register_extcmd ("date", grub_cmd_date,
+			  GRUB_COMMAND_ACCEPT_DASH
+			  | GRUB_COMMAND_OPTIONS_AT_START,
+			  N_("-n | [[year-]month-day] hour:minute[:second]]"),
+			  N_("Display/set current datetime."),
+			  options);
 }
 
 GRUB_MOD_FINI(date)
 {
-  grub_unregister_command (cmd);
+  grub_unregister_extcmd (cmd);
 }
