diff --git a/src/luks/clevis-luks-common-functions.in b/src/luks/clevis-luks-common-functions.in
index 29e4631..0f335b8 100644
--- a/src/luks/clevis-luks-common-functions.in
+++ b/src/luks/clevis-luks-common-functions.in
@@ -810,7 +810,7 @@ clevis_luks_add_key() {
         extra_args="$(printf -- '--token-id %s' "${EXISTING_TOKEN_ID}")"
         input="$(printf '%s' "${NEWKEY}")"
     fi
-    local pbkdf_args="--pbkdf pbkdf2 --pbkdf-force-iterations 1000"
+    local pbkdf_args="-h sha512"
 
     printf '%s' "${input}" | cryptsetup luksAddKey --force-password --batch-mode \
                                          --key-slot "${SLT}" \
@@ -850,7 +850,7 @@ clevis_luks_update_key() {
         input="$(printf '%s' "${NEWKEY}")"
     fi
 
-    local pbkdf_args="--pbkdf pbkdf2 --pbkdf-force-iterations 1000"
+    local pbkdf_args="-h sha512"
 
     if [ -n "${in_place}" ]; then
         printf '%s' "${input}" | cryptsetup luksChangeKey "${DEV}" \
diff --git a/src/luks/tests/tests-common-functions.in b/src/luks/tests/tests-common-functions.in
index bfd7547..6016033 100755
--- a/src/luks/tests/tests-common-functions.in
+++ b/src/luks/tests/tests-common-functions.in
@@ -69,8 +69,7 @@ new_device() {
     fi
 
     fallocate -l64M "${DEV}"
-    cryptsetup luksFormat --type "${LUKS}" --pbkdf pbkdf2 \
-        --pbkdf-force-iterations 1000 --key-size 512 --batch-mode \
+    cryptsetup luksFormat --type "${LUKS}" -h sha512 --key-size 512 --batch-mode \
         --force-password "${DEV}" <<< "${PASS}"
     # Caching the just-formatted device for possible reuse.
     cp -f "${DEV}" "${DEV_CACHED}"
@@ -93,8 +92,7 @@ new_device_keyfile() {
     fi
 
     fallocate -l64M "${DEV}"
-    cryptsetup luksFormat --type "${LUKS}" --pbkdf pbkdf2 \
-        --pbkdf-force-iterations 1000 --key-size 512 --batch-mode \
+    cryptsetup luksFormat --type "${LUKS}" -h sha512 --key-size 512 --batch-mode \
         "${DEV}" "${KEYFILE}"
 }
 
